sql/tests: TestRandomSyntaxSQLSmith failed [trying to apply inverted index to unsupported type NAME]

sql/tests.TestRandomSyntaxSQLSmith [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/13526051?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/13526051?buildTab=artifacts#/) on release-23.2 @ [833ddb62e2773d1bc962cdb77a182bbe8b8b2f2f](https://github.com/cockroachdb/cockroach/commits/833ddb62e2773d1bc962cdb77a182bbe8b8b2f2f):

Random syntax error:

```
    rsg_test.go:911: Crash detected: server panic: pq: internal error: trying to apply inverted index to unsupported type NAME
```
Query:

```
        SELECT
        	tab378595."c""ol1_
11" AS "col""897895",
        	tab378595."col\\xb51_3" AS col897896,
        	ARRAY['778941CA/1FEBB88F':::PG_LSN,'D32CE40E/A104D09F':::PG_LSN] AS col897897,
        	tab378595.col1_7 AS ".col 897898",
        	(-8.207504212976660413E+33):::DECIMAL AS col897899,
        	NULL AS col897900,
        	tab378598.col897893 AS " %pco""l897901",
        	'9{~)/T]':::STRING AS "c%vo L897902",
        	NULL AS col897903,
        	B'01001010111000001110' AS "c%po""l897904",
        	tab378595."co😭l 1_5" AS "c.ol'😬897905",
        	e'c\x0e0\x13;wT7\'':::STRING:::NAME AS "col 897906",
        	tab378595."co😭l 1_5" AS còl897907,
        	'BOX(-0.21275307529907383 -0.5033230890123566,1.5648962163983002 1.724787000786968)':::BOX2D
        		AS "c\\u7C3Eol̩897908",
        	e'w\x0b':::STRING:::NAME AS col897909,
        	tab378595.tableoid AS "col%p897910"
        FROM
        	defaultdb.public.table1@[0] AS tab378595
        	JOIN (
        			SELECT
        				'~Z(k':::STRING:::NAME AS col897891, name(B'001010000110110011000111010101110'::BIT)::NAME AS col897892
        			FROM
        				defaultdb.public.table1 AS tab378596
        				JOIN defaultdb.public.table1 AS "ta b378597" ON (tab378596.col1_10) = ("ta b378597".col1_10)
        			ORDER BY
        				tab378596.col1_10 ASC NULLS FIRST
        			LIMIT
        				45:::INT8
        		)
        			AS tab378598 (col897893, " col897894") ON
        			(tab378595."col%v1_2") = (tab378598.col897893) AND (tab378595.col1_10) = (tab378598." col897894")
        ORDER BY
        	tab378598." col897894" ASC NULLS LAST,
        	tab378595."col\\xb51_3" DESC NULLS LAST,
        	tab378595.crdb_internal_mvcc_timestamp DESC NULLS LAST,
        	tab378595.col1_10 NULLS FIRST,
        	tab378595.c😯ol1_1 DESC NULLS FIRST
        LIMIT
        	5:::INT8;
```
Schema:

```
    rsg_test.go:720: To reproduce, use schema:
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.automatic_collection.enabled = false;;
    rsg_test.go:722: SET CLUSTER SETTING sql.stats.histogram_collection.enabled = false;;
    rsg_test.go:722: CREATE TABLE table1 ("c-ol1_0" DECIMAL NOT NULL, c😯ol1_1 BIT(20) NOT NULL, "col%v1_2" VARCHAR NOT NULL, "col\\xb51_3" DATE NOT NULL, col1_4 PG_LSN[] NOT NULL, "co😭l 1_5" REGTYPE, "co-l1_6" TSQUERY NULL, col1_7 TIMESTAMPTZ NOT NULL, "co%vl1_8" BOX2D, col1_9 JSONB NOT NULL, col1_10 NAME NOT NULL, "c""ol1_
11" GEOGRAPHY, "col 1_12" REGCLASS NULL, UNIQUE (col1_4) WHERE (((table1."col\\xb51_3" != '-infinity':::DATE) AND (table1."c-ol1_0" >= (-1):::DECIMAL)) OR (table1.col1_10 < e'\x00':::STRING:::NAME)) AND (table1.col1_7 <= '0001-01-01 00:00:00+00':::TIMESTAMPTZ), INDEX ("co%vl1_8" ASC) STORING (c😯ol1_1, "col%v1_2", "col\\xb51_3", col1_4, "co😭l 1_5", "col 1_12") WHERE table1.col1_10 > e'\U00002603':::STRING:::NAME VISIBILITY 0.95, UNIQUE (col1_10 DESC, "col 1_12", "c-ol1_0"), INVERTED INDEX (col1_10 gin_trgm_ops) WHERE ((((table1."col\\xb51_3" >= 'infinity':::DATE) OR (table1."c-ol1_0" = 1:::DECIMAL)) OR (table1.col1_7 >= '2001-01-01 00:00:00+00 BC':::TIMESTAMPTZ)) OR (table1."col%v1_2" != 'X':::STRING)) OR (table1.col1_10 != 'X':::STRING:::NAME), FAMILY (col1_9, "c-ol1_0", "c""ol1_
11", col1_7), FAMILY (col1_4), FAMILY ("col\\xb51_3", col1_10), FAMILY ("col%v1_2"), FAMILY (c😯ol1_1, "col 1_12"), FAMILY ("co-l1_6", "co😭l 1_5"), FAMILY ("co%vl1_8"));
    rsg_test.go:722: CREATE TABLE table2 ("cO'l2_0" INT8 NULL, col2_1 REFCURSOR, col2_2 JSONB NOT NULL, col2_3 STRING COLLATE de_DE NOT NULL, "col2'_4" STRING COLLATE de_DE NULL, col2_5 BIT(33) NOT NULL, "col\\x422_6" VARCHAR[], col2_7 NAME NOT NULL, "col*2_8" OID NULL, "c)ol2_\\xa49" REGROLE, col2_10 STRING NOT NULL AS (lower(col2_7)) STORED, INDEX ("col\\x422_6" DESC) STORING (col2_2, "col2'_4", col2_5, col2_7, "c)ol2_\\xa49") WHERE ((table2."cO'l2_0" > (-9223372036854775807):::INT8) OR (table2.col2_7 = '"':::STRING:::NAME)) OR (table2.col2_10 < e'\x00':::STRING), INDEX (col2_5 DESC), UNIQUE ("cO'l2_0" DESC, "col*2_8" DESC), UNIQUE ("col2'_4" DESC) STORING ("cO'l2_0", col2_1, col2_2, col2_3, col2_7, "col*2_8") WHERE ((table2.col2_7 >= '':::STRING:::NAME) AND (table2."cO'l2_0" < 32767:::INT8)) AND (table2.col2_10 < e'\'':::STRING), UNIQUE ("col\\x422_6"), INVERTED INDEX ("col\\x422_6", col2_10 gin_trgm_ops), FAMILY (col2_5), FAMILY (col2_1), FAMILY ("cO'l2_0", "c)ol2_\\xa49", "col2'_4", col2_7, "col\\x422_6", "col*2_8"), FAMILY (col2_3), FAMILY (col2_2), FAMILY (col2_10));
    rsg_test.go:722: ALTER TABLE table1 INJECT STATISTICS e'[{"avg_size": 0, "columns": ["\\"co-l1_6\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["col1_9"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["c\U0001F62Fol1_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["\\"col%v1_2\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["\\"col\\\\\\\\xb51_3\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["\\"co%vl1_8\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_buckets": [{"distinct_range": 0, "num_eq": 10000000000, "num_range": 0, "upper_bound": "BOX(-2.602909955714282 0.17268414301408797,-1.156766336373008 0.7035990813749418)"}, {"distinct_range": 125.89580636483527, "num_eq": 8000000000, "num_range": 200, "upper_bound": "BOX(-1.7423396230082844 -0.46243049847645223,-0.2092810136662676 0.11708808351923086)"}, {"distinct_range": 3000000000, "num_eq": 600, "num_range": 3000000000, "upper_bound": "BOX(-0.7411503104888769 -0.08589435141914536,-0.6319036039038348 0.5285670965067666)"}, {"distinct_range": 0, "num_eq": 5162420332362452513, "num_range": 10000, "upper_bound": "BOX(0.06346724224170958 -0.19712133098841284,0.10567955421274505 0.01870452486573604)"}], "histo_col_type": "BOX2D", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["\\"col 1_12\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["\\"co\U0001F62Dl 1_5\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["col1_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["\\"c\\"\\"ol1_\\r11\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["\\"c-ol1_0\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["col1_4"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_buckets": [{"distinct_range": 0, "num_eq": 7054329270108900031, "num_range": 0, "upper_bound": "ARRAY[\'8CED7692/1076CD56\':::PG_LSN,\'1128AA6C/22080617\':::PG_LSN,\'329967D0/136BDD7A\':::PG_LSN]"}, {"distinct_range": 0, "num_eq": 5885459844043241887, "num_range": 6645525466179504941, "upper_bound": "ARRAY[\'9C61172A/E3C4E6DA\':::PG_LSN,\'127F2075/36D67774\':::PG_LSN,\'8387185/3C5C1\':::PG_LSN,\'4D479EBB/7899E17A\':::PG_LSN]"}, {"distinct_range": 588.207774301771, "num_eq": 4131456377728882584, "num_range": 600, "upper_bound": "ARRAY[\'A1B48193/632C61F2\':::PG_LSN,\'99C6CE77/401ED0FD\':::PG_LSN,\'E13477B/E1DAF452\':::PG_LSN,\'CE9505BF/163AFCC5\':::PG_LSN]"}], "histo_col_type": "PG_LSN[]", "name": "__auto__", "null_count": 0, "row_count": 0}, {"avg_size": 0, "columns": ["col1_10"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 0, "histo_buckets": [{"distinct_range": 0, "num_eq": 5683041247861354162, "num_range": 0, "upper_bound": "\\ns#\\u000fbF?T"}, {"distinct_range": 7399730802916391000, "num_eq": 4158559004407306809, "num_range": 7399730802916390839, "upper_bound": "\\"\\u000f;J\\tbI#"}, {"distinct_range": 96683711.9276624, "num_eq": 1844749348370758898, "num_range": 100000000, "upper_bound": "X"}, {"distinct_range": 4256343037145817600, "num_eq": 2458358276822925710, "num_range": 4256343037145817411, "upper_bound": "]6g\\u00087"}, {"distinct_range": 0, "num_eq": 7000000000, "num_range": 900, "upper_bound": "p%\\u0014\\u0003b"}, {"distinct_range": 0, "num_eq": 4416725743177325921, "num_range": 4000000000, "upper_bound": "\U00002603"}], "histo_col_type": "NAME", "name": "__auto__", "null_count": 0, "row_count": 0}]':::JSONB;
    rsg_test.go:722: ALTER TABLE table2 INJECT STATISTICS e'[{"avg_size": 23, "columns": ["col2_2"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2677349273, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8000000000}, {"avg_size": 0, "columns": ["col2_3"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1091392497, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8000000000}, {"avg_size": 3, "columns": ["\\"col2\'_4\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 7771179924, "histo_col_type": "", "name": "__auto__", "null_count": 9040720, "row_count": 8000000000}, {"avg_size": 30, "columns": ["col2_5"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 1885842232, "histo_buckets": [{"distinct_range": 0, "num_eq": 500000, "num_range": 0, "upper_bound": "000010110101010011111101010101110"}, {"distinct_range": 400000, "num_eq": 4117284556956923528, "num_range": 400000, "upper_bound": "011010100111100111010100101101010"}, {"distinct_range": 500, "num_eq": 70000000000, "num_range": 500, "upper_bound": "101011101011011000011100010001011"}, {"distinct_range": 832467436293939300, "num_eq": 0, "num_range": 5958174094945719022, "upper_bound": "110000010000110011000010011010111"}, {"distinct_range": 0, "num_eq": 100000, "num_range": 2866919338201274581, "upper_bound": "110011010011110001010110110110011"}, {"distinct_range": 0, "num_eq": 2325504121976712473, "num_range": 50000000, "upper_bound": "110101000000000111011000001111000"}], "histo_col_type": "BIT(33)", "name": "__auto__", "null_count": 0, "row_count": 8000000000}, {"avg_size": 17, "columns": ["col2_7"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 7381365470, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8000000000}, {"avg_size": 19, "columns": ["\\"cO\'l2_0\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 3823421773, "histo_buckets": [{"distinct_range": 0, "num_eq": 90000000, "num_range": 0, "upper_bound": "-8669478491521847803"}, {"distinct_range": 0, "num_eq": 10, "num_range": 0, "upper_bound": "-6959055455667755578"}, {"distinct_range": 0, "num_eq": 80000000000, "num_range": 3265392725121744613, "upper_bound": "-5609564428346213229"}, {"distinct_range": 0, "num_eq": 730506587401263301, "num_range": 2024831432391028388, "upper_bound": "-3728314042525190408"}, {"distinct_range": 1927926074854194000, "num_eq": 10000000000, "num_range": 7537702332406249866, "upper_bound": "964664744101496522"}, {"distinct_range": 300000, "num_eq": 328606170742854970, "num_range": 300000, "upper_bound": "6465856783419518530"}], "histo_col_type": "INT8", "name": "__auto__", "null_count": 1743723528, "row_count": 8000000000}, {"avg_size": 12, "columns": ["col2_1"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 2635429757, "histo_col_type": "", "name": "__auto__", "null_count": 4695075604, "row_count": 8000000000}, {"avg_size": 1, "columns": ["\\"col\\\\\\\\x422_6\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 478921018, "histo_buckets": [{"distinct_range": 0, "num_eq": 200000, "num_range": 0, "upper_bound": "ARRAY[\'Yh\':::STRING]"}, {"distinct_range": 0, "num_eq": 900000, "num_range": 0, "upper_bound": "ARRAY[e\'z\\\\x7f\':::STRING,e\'v?9\\\\x1cAeS^l\':::STRING]"}, {"distinct_range": 95.53405015272065, "num_eq": 2476166711738756971, "num_range": 100, "upper_bound": "ARRAY[e\'\\\\U00002603\':::STRING]"}], "histo_col_type": "VARCHAR[]", "name": "__auto__", "null_count": 1479267500, "row_count": 8000000000}, {"avg_size": 16, "columns": ["\\"col*2_8\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5749524745, "histo_col_type": "", "name": "__auto__", "null_count": 3730616427, "row_count": 8000000000}, {"avg_size": 12, "columns": ["\\"c)ol2_\\\\\\\\xa49\\""], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 175609470, "histo_col_type": "", "name": "__auto__", "null_count": 3391075808, "row_count": 8000000000}, {"avg_size": 25, "columns": ["col2_10"], "created_at": "2000-01-01 00:00:00+00:00", "distinct_count": 5367340105, "histo_col_type": "", "name": "__auto__", "null_count": 0, "row_count": 8000000000}]':::JSONB;
    rsg_test.go:722: CREATE TYPE rand_typ_0 AS ENUM ('qlsr', 'fluye', 'oecw', 'oottfg', 'dnuyo', 'nbgx');
    rsg_test.go:722: CREATE TYPE rand_typ_1 AS ENUM ('kh', 'qsvsh', 'fo', 'rsdxk', 'dybcf', 'men');
    rsg_test.go:722: SET sql_safe_updates = false;;
    rsg_test.go:722: 
        BEGIN; CREATE TYPE greeting AS ENUM ('hello', 'howdy', 'hi', 'good day', 'morning'); COMMIT;
        BEGIN;
        CREATE TABLE IF NOT EXISTS seed AS
        	SELECT
        		g::INT2 AS _int2,
        		g::INT4 AS _int4,
        		g::INT8 AS _int8,
        		g::FLOAT4 AS _float4,
        		g::FLOAT8 AS _float8,
        		'2001-01-01'::DATE + g AS _date,
        		'2001-01-01'::TIMESTAMP + g * '1 day'::INTERVAL AS _timestamp,
        		'2001-01-01'::TIMESTAMPTZ + g * '1 day'::INTERVAL AS _timestamptz,
        		g * '1 day'::INTERVAL AS _interval,
        		g % 2 = 1 AS _bool,
        		g::DECIMAL AS _decimal,
        		g::STRING AS _string,
        		g::STRING::BYTES AS _bytes,
        		substring('00000000-0000-0000-0000-' || g::STRING || '00000000000', 1, 36)::UUID AS _uuid,
        		'0.0.0.0'::INET + g AS _inet,
        		g::STRING::JSONB AS _jsonb,
        		enum_range('hello'::greeting)[g] as _enum
        	FROM
        		generate_series(1, 5) AS g;
        COMMIT;
        
        INSERT INTO seed DEFAULT VALUES;
        CREATE INDEX on seed (_int8, _float8, _date);
        CREATE INVERTED INDEX on seed (_jsonb);
        ;
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (857474738045732253);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (6612317592417023808);
    rsg_test.go:722: ALTER TABLE seed SPLIT AT VALUES (6435324310175971997);
    rsg_test.go:722: ALTER TABLE seed SCATTER;
    rsg_test.go:724: 
    rsg_test.go:725: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/d437d2c847dfedbc4972f231c3331c8e/logTestRandomSyntaxSQLSmith573670146
--- FAIL: TestRandomSyntaxSQLSmith (385.35s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxSQLSmith.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-35336

--- a/pkg/sql/logictest/testdata/logic_test/trigram_indexes
+++ b/pkg/sql/logictest/testdata/logic_test/trigram_indexes
@@ -335,3 +335,26 @@ CREATE TABLE public.t_112713 (
 )
 
 subtest end
+
+# Regression test for hitting an internal error on string-like datums (e.g NAME).
+statement ok
+CREATE TABLE t117758 (
+  col1 VARCHAR NOT NULL,
+  col2 NAME NOT NULL,
+  INVERTED INDEX (col2 gin_trgm_ops)
+);
+SELECT
+    tab.col1_1
+FROM
+    t117758 AS tab2
+    JOIN (
+            SELECT
+                'foo'::NAME, 'bar'::NAME
+            FROM
+                t117758 AS tab3
+                JOIN t117758 AS tab4 ON
+                        (tab3.col2) = (tab4.col2)
+        )
+            AS tab (col1_1, col1_2) ON
+            (tab2.col1) = (tab.col1_1)
+            AND (tab2.col2) = (tab.col1_2);

--- a/pkg/sql/opt/invertedidx/trigram.go
+++ b/pkg/sql/opt/invertedidx/trigram.go
@@ -21,7 +21,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
-	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/errors"
 )
 
@@ -95,13 +94,14 @@ func (t *trigramFilterPlanner) extractInvertedFilterConditionFromLeaf(
 		// Can only accelerate with a single constant value.
 		return inverted.NonInvertedColExpression{}, expr, nil
 	}
-	d := memo.ExtractConstDatum(constantVal)
-	if d.ResolvedType() != types.String {
+	d := tree.UnwrapDOidWrapper(memo.ExtractConstDatum(constantVal))
+	ds, ok := d.(*tree.DString)
+	if !ok {
 		panic(errors.AssertionFailedf(
 			"trying to apply inverted index to unsupported type %s", d.ResolvedType().SQLStringForError(),
 		))
 	}
-	s := string(*d.(*tree.DString))
+	s := string(*ds)
 	var err error
 	invertedExpr, err = rowenc.EncodeTrigramSpans(s, allMustMatch)
 	if err != nil {

